### PR TITLE
Dryer lint can be good

### LINKED ISSUE #65: Introduce go linting to CI process
* **What are you trying to do?**

The issue that spurred PR #59 was found via running a go linter. https://github.com/golangci/golangci-lint to be specific. But we should get this or at a dead minimum `go vet` in place as part of our CI pipeline. This will help our code suck less. 

* **What feature or behavior is this required for?**
Doing linting on all PRs and master builds

* **How could we solve this issue? (Not knowing is okay!)**
Introduce `golangci-lint` or `go vet` as part of the build process. 

* **Anything else?**
`golangci-lint` has a lot of configurability as well as lots more linters/tools that come with that have the ability to help find and fix issues. Whereas `go vet` is a do one thing and only that thing. It did point out the same issue fixed as part of #59 but that was it. 

`golangci-lint` does in fact run `go vet` as part of its process. 

The defaults for `golangci-lint` are as follows : 
image

@flimzy mentioned that he likes to configure some others that are not enabled by default so hopefully they will fill us in on those details and we can discuss further. 

I currently dont have strong opinions on what we enable at the moment. At this point getting either in place or even a different option is better than the nothing we have today.

cc @bhamail / @DarthHater / @flimzy

### PR DESCRIPTION
**Pull request by zendern**
Adding golangci-lint to the project as well as start to use Make to encapsulate all build things. 

It relates to the following issue #s:
* Fixes #65 

cc @bhamail / @DarthHater / @flimzy

### CODE REVIEW

**Comment by zendern** [comment]
Ok commit 827ae25 is what i would expect. It now fails on linting and stops at that step when it does. Once #59 is merged those issues will be dealt and this build should be green post rebase. 

https://travis-ci.org/sonatype-nexus-community/nancy/builds/639931932?utm_source=github_status&utm_medium=notification

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	$(GOCMD) mod tidy
+
+env-setup:
+	export GO111MODULE=on CGO_ENABLED=0
+
+build: env-setup
+	$(GOBUILD) -o $(BINARY_NAME) -v
+
+test:
+	$(GOTEST) -v ./...
```
Comment (flimzy) [inline Makefile]: This may not really relate to this PR, but running tests with the race detector (i.e. `go test -race -v ./...`) is often a good idea, too, and is at least tangentially related to linting :)

Comment (zendern) [reply Makefile]: race

:) im up for `racing` ... added in commit c4fbd81

**Comment by DarthHater** [comment]
The title of this PR just made sense to me (my dryer broke) and I'm DYING

**Comment by zendern** [comment]
So this PR is actually all good minus the fact that some of the new tests have race conditions. I'll take a stab at them hopefully this week or if @DarthHater you'd like to take an 👀 since they are the new Nexus iQ tests. 

```
Failure - iq
Failed
==================
WARNING: DATA RACE
Read at 0x00000107ee68 by goroutine 121:
  github.com/sonatype-nexus-community/nancy/iq.AuditPackages.func1()
      /go/src/github.com/sonatype-nexus-community/nancy/iq/iq.go:97 +0x43

Previous write at 0x00000107ee68 by goroutine 114:
  github.com/sonatype-nexus-community/nancy/iq.AuditPackages()
      /go/src/github.com/sonatype-nexus-community/nancy/iq/iq.go:67 +0x7f
  github.com/sonatype-nexus-community/nancy/iq.TestAuditPackagesIqUpButBadThirdPartyAPIResponse()
      /go/src/github.com/sonatype-nexus-community/nancy/iq/iq_test.go:169 +0x5d0
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:909 +0x199

Goroutine 121 (running) created at:
  github.com/sonatype-nexus-community/nancy/iq.AuditPackages()
      /go/src/github.com/sonatype-nexus-community/nancy/iq/iq.go:91 +0x2d6
  github.com/sonatype-nexus-community/nancy/iq.TestAuditPackages()
      /go/src/github.com/sonatype-nexus-community/nancy/iq/iq_test.go:99 +0x67c
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:909 +0x199

Goroutine 114 (finished) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:960 +0x651
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1202 +0xa6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:909 +0x199
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1200 +0x521
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1117 +0x2ff
  main.main()
      _testmain.go:48 +0x223
==================
==================
WARNING: DATA RACE
Read at 0x00000107ee78 by goroutine 121:
  github.com/sonatype-nexus-community/nancy/iq.AuditPackages.func1()
      /go/src/github.com/sonatype-nexus-community/nancy/iq/iq.go:97 +0x112

Previous write at 0x00000107ee78 by goroutine 114:
  github.com/sonatype-nexus-community/nancy/iq.AuditPackages()
      /go/src/github.com/sonatype-nexus-community/nancy/iq/iq.go:67 +0x7f
  github.com/sonatype-nexus-community/nancy/iq.TestAuditPackagesIqUpButBadThirdPartyAPIResponse()
      /go/src/github.com/sonatype-nexus-community/nancy/iq/iq_test.go:169 +0x5d0
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:909 +0x199

Goroutine 121 (running) created at:
  github.com/sonatype-nexus-community/nancy/iq.AuditPackages()
      /go/src/github.com/sonatype-nexus-community/nancy/iq/iq.go:91 +0x2d6
  github.com/sonatype-nexus-community/nancy/iq.TestAuditPackages()
      /go/src/github.com/sonatype-nexus-community/nancy/iq/iq_test.go:99 +0x67c
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:909 +0x199

Goroutine 114 (finished) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:960 +0x651
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1202 +0xa6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:909 +0x199
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1200 +0x521
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1117 +0x2ff
  main.main()
      _testmain.go:48 +0x223
==================
Found 2 data race(s)```

**Comment by zendern** [comment]
@bhamail  @DarthHater  @flimzy @fitzoh 

This PR is now back in order after the flurry of PRs we have had recently. 

To recap : 
1. gets `golangci-lint` in place. Locally i have it weird up to use docker cause that felt the most reproducible to make everyone run on the version we want of golangci-lint. On ci to not deal with docker in docker i just install it directly but using the same version. `make ci-lint`. So you could also run this locally ¯\_(ツ)_/¯ 
2. fixes all linting issues
3. starts use of makefile to make the ci and local build, test, lint things as easy as `make lint`, `make test`, `make build`, `make integration-test` instead of needing to figure out what to run exactly to do the things

Earlier in this PR @flimzy pointed out we should add `-race` flag to our tests. I did...it broke stuff once i rebased :) which is good but I'm going to create an issue to deal with that separately, this PR already feels big. Errors when running with `-race` flag are coming from the new iq stuff cc @DarthHater .

```
==================
WARNING: DATA RACE
Read at 0x000001ca01a8 by goroutine 43:
  github.com/sonatype-nexus-community/nancy/iq.AuditPackages.func1()
      /Users/zender/dev/projects/oss/nancy/iq/iq.go:112 +0x43

Previous write at 0x000001ca01a8 by goroutine 77:
  github.com/sonatype-nexus-community/nancy/iq.AuditPackages()
      /Users/zender/dev/projects/oss/nancy/iq/iq.go:73 +0x2b0
  github.com/sonatype-nexus-community/nancy/iq.TestAuditPackagesIqUpButBadThirdPartyAPIResponse()
      /Users/zender/dev/projects/oss/nancy/iq/iq_test.go:169 +0x5f9
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:909 +0x199

Goroutine 43 (running) created at:
  github.com/sonatype-nexus-community/nancy/iq.AuditPackages()
      /Users/zender/dev/projects/oss/nancy/iq/iq.go:106 +0x808
  github.com/sonatype-nexus-community/nancy/iq.TestAuditPackages()
      /Users/zender/dev/projects/oss/nancy/iq/iq_test.go:99 +0x6a5
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:909 +0x199

Goroutine 77 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:960 +0x651
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1202 +0xa6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:909 +0x199
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1200 +0x521
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1117 +0x2ff
  main.main()
      _testmain.go:48 +0x223
==================
==================
WARNING: DATA RACE
Read at 0x000001ca01b8 by goroutine 43:
  github.com/sonatype-nexus-community/nancy/iq.AuditPackages.func1()
      /Users/zender/dev/projects/oss/nancy/iq/iq.go:112 +0x112

Previous write at 0x000001ca01b8 by goroutine 77:
  github.com/sonatype-nexus-community/nancy/iq.AuditPackages()
      /Users/zender/dev/projects/oss/nancy/iq/iq.go:73 +0x2b0
  github.com/sonatype-nexus-community/nancy/iq.TestAuditPackagesIqUpButBadThirdPartyAPIResponse()
      /Users/zender/dev/projects/oss/nancy/iq/iq_test.go:169 +0x5f9
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:909 +0x199

Goroutine 43 (running) created at:
  github.com/sonatype-nexus-community/nancy/iq.AuditPackages()
      /Users/zender/dev/projects/oss/nancy/iq/iq.go:106 +0x808
  github.com/sonatype-nexus-community/nancy/iq.TestAuditPackages()
      /Users/zender/dev/projects/oss/nancy/iq/iq_test.go:99 +0x6a5
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:909 +0x199

Goroutine 77 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:960 +0x651
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1202 +0xa6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:909 +0x199
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1200 +0x521
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1117 +0x2ff
  main.main()
      _testmain.go:48 +0x223
==================
--- FAIL: TestAuditPackagesIqUpButBadThirdPartyAPIResponse (1.11s)
    testing.go:853: race detected during execution of test
FAIL
FAIL	github.com/sonatype-nexus-community/nancy/iq	2.094s
```

File: configuration/parse.go [hunk 2]
```diff
@@ -209,7 +208,7 @@ func determineIfLineIsExclusion(ogLine string, config *Configuration) error {
 		if until != nil {
 			parseDate, err := time.Parse("2006-01-02", strings.TrimSpace(until[2]))
 			if err != nil {
-				return errors.New(fmt.Sprintf("failed to parse until at line '%s'. Expected format is 'until=yyyy-MM-dd'", ogLine))
+				return fmt.Errorf("failed to parse until at line '%s'. Expected format is 'until=yyyy-MM-dd'", ogLine)
```
Comment (greut) [inline configuration/parse.go]:
```suggestion
				return fmt.Errorf("failed to parse until at line %q. Expected format is 'until=yyyy-MM-dd'", ogLine)
```

File: packages/mod.go [hunk 3]
```diff
@@ -30,8 +30,7 @@ type Mod struct {
 
 func (m Mod) ExtractPurlsFromManifest() (purls []string) {
 	for _, s := range m.ProjectList.Projects {
-		var version string
-		version = strings.Replace(s.Version, "v", "", -1)
+		version := strings.Replace(s.Version, "v", "", -1)
```
Comment (greut) [inline packages/mod.go]: Not part of this PR, but it could be nice to use a library for this, https://pkg.go.dev/golang.org/x/mod/semver

File: configuration/parse.go (hunk 2 above)
Comment (zendern) [reply configuration/parse.go]: Applied in 57dc506

File: packages/mod.go (hunk 3 above)
Comment (zendern) [reply packages/mod.go]: @greut Im not sure I see  how that library helps here....from what i can tell it will still always return a `v` on the version number. We also use this library as well. https://github.com/Masterminds/semver. That might be an option too but would still requires  recombining the MAJOR, MINOR and PATCH numbers it parsed back into something we want just to get rid of the `v` that may or may not be there.

Comment (greut) [reply packages/mod.go]: @zendern I see! thanks for the clarification.

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+	rm -f $(BINARY_NAME)
+	rm -f $(BINARY_NAME)-*amd64*
+
+deps:
+	$(GOCMD) mod download
+	$(GOCMD) mod verify
+	$(GOCMD) mod tidy
+
+env-setup:
+	export GO111MODULE=on CGO_ENABLED=0
```
Comment (fitzoh) [inline Makefile]: Does this work?
I didn't think exported vars persisted to other steps.


Testing with this Makefile leads me to believe it doesn't
```
stuff:
        export WHATEVER=blah

things: stuff
        env
        echo $(WHATEVER)
```

File: logger/logger.go [hunk 5]
```diff
@@ -80,7 +81,10 @@ func useTestLogFile(args []string) bool {
 // GetLogFileLocation will return the location on disk of the log file
 func GetLogFileLocation() (result string) {
 	result, _ = os.UserHomeDir()
-	os.MkdirAll(path.Join(result, types.OssIndexDirName), os.ModePerm)
+	err := os.MkdirAll(path.Join(result, types.OssIndexDirName), os.ModePerm)
+	if err != nil {
+		fmt.Print(err)
```
Comment (fitzoh) [inline logger/logger.go]: is that error going to make sense in isolation?

File: Makefile (hunk 4 above)
Comment (zendern) [reply Makefile]: Fixed here 3421af6

File: logger/logger.go (hunk 5 above)
Comment (zendern) [reply logger/logger.go]: Yeah it wasn't great...followed the other error print in here. So I now fixed them both in 124d008

File: Makefile (hunk 4 above)
Comment (fitzoh) [reply Makefile]: I liked the idea.

Might make sense to make `GO_BUILD_FLAGS` a top level var if you want to keep the explicitness.... Or just :shipit:

Comment (zendern) [reply Makefile]: Done b4b18fb

**Comment by bhamail** [comment]
Hoi hoi. So I've become a lint fan boi. I resolved the conflicts. The linter is doing it's job, and found some more stuff to fix. I'm gonna merge this PR, and then do the linter cleanups on master. Duck and cover.

**Comment by zendern** [comment]
> Hoi hoi. So I've become a lint fan boi. I resolved the conflicts. The linter is doing it's job, and found some more stuff to fix. I'm gonna merge this PR, and then do the linter cleanups on master. Duck and cover. 

🎉🎉🎉🎉 🦆🦆💥💥💥
